File: ios/LivenessDetection/LivenessConfig.swift
import Foundation
import UIKit

@objc public final class LivenessConfig: NSObject {
  @objc public var readyMs: Int64 = 1800
  @objc public var sessionTimeoutMs: Int64 = 120000
  @objc public var baselineFrames: Int = 8
  @objc public var yawTurnDelta: Float = 9
  @objc public var yawWrongDirDelta: Float = 16
  @objc public var headTurnHoldMs: Int64 = 80
  @objc public var nodDownDelta: Float = 3
  @objc public var nodReturnFraction: Float = 0.85
  @objc public var nodReturnMaxDelta: Float = 12
  @objc public var blinkClosedThreshold: Float = 0.30
  @objc public var blinkOpenThreshold: Float = 0.25
  @objc public var earClosedThreshold: Float = 0.20
  @objc public var earOpenThreshold: Float = 0.25
  @objc public var blinkMaxDurationMs: Int64 = 4000
  @objc public var mouthOpenThreshold: Float = 0.20
  @objc public var mouthOpenMarThreshold: Float = 0.20
  @objc public var mouthHoldMs: Int64 = 50
  @objc public var maxYawDuringBlink: Float = 30
  @objc public var maxPitchDuringBlink: Float = 30
  @objc public var maxYawDuringNod: Float = 40
  @objc public var maxYawDuringMouth: Float = 35
  @objc public var maxPitchDuringMouth: Float = 35
  @objc public var ovalCx: Float = 0.50
  @objc public var ovalCy: Float = 0.42
  @objc public var ovalRx: Float = 0.32
  @objc public var ovalRy: Float = 0.40
  @objc public var minFaceSize: Float = 0.10
  @objc public var maxFaceSize: Float = 0.62
  @objc public var captureDelayMs: Int64 = 700
  @objc public var captureMaxAttempts: Int = 90
  @objc public var captureMaxYaw: Float = 18
  @objc public var captureMaxPitch: Float = 18
  @objc public var captureMaxMouthScore: Float = 0.20
  @objc public var captureMaxBlinkScore: Float = 0.25
  @objc public var captureMinEar: Float = 0.22
  @objc public var captureMaxMar: Float = 0.22
  @objc public var shuffleSteps: Bool = true

  @objc public var faceDetectionConfidence: Float = 0.5
  @objc public var facePresenceConfidence: Float = 0.5
  @objc public var faceTrackingConfidence: Float = 0.5

  @objc public var cdnMaxRetries: Int = 5
  @objc public var cdnAttemptTimeoutMs: Int64 = 45000
  @objc public var connectivityCheckTimeoutMs: Int64 = 5000

  // ── UI: shape, theme, layout (mirrors @daboss2003/liveness-web options) ───
  /// "oval" (default) or "circle".
  @objc public var shape: String = "oval"
  /// When false, SDK-rendered instruction text, position hint, gesture icon and step dots are hidden. Sounds still play.
  @objc public var showInstructions: Bool = true
  /// Minimum diameter (pt) of the visible face frame so it stays usable on small screens. 0 = no floor.
  @objc public var minSize: CGFloat = 220
  /// Ring stroke colour while face is in position.
  @objc public var progressColor: UIColor = UIColor(red: 18/255, green: 201/255, blue: 92/255, alpha: 1)
  /// Ring stroke colour while face is out of position.
  @objc public var progressErrorColor: UIColor = UIColor(red: 1, green: 59/255, blue: 59/255, alpha: 1)
  /// Ring stroke width in pt.
  @objc public var progressWidth: CGFloat = 3.5
  /// "round", "square" or "butt".
  @objc public var progressLineCap: String = "round"
  /// Mask fill colour outside the cutout while face is in position.
  @objc public var overlayColor: UIColor = UIColor(white: 0, alpha: 0.82)
  /// Mask fill colour outside the cutout while face is out of position.
  @objc public var overlayErrorColor: UIColor = UIColor(red: 180/255, green: 0, blue: 0, alpha: 0.55)

  @objc public override init() {
    super.init()
  }
}

File: android/liveness/src/main/java/com/liveness/detection/LivenessConfig.kt
package com.liveness.detection

import android.graphics.Color

data class LivenessConfig(
  val readyMs: Long = 1800L,
  val sessionTimeoutMs: Long = 120000L,
  val baselineFrames: Int = 8,
  val yawTurnDelta: Float = 9f,
  val yawWrongDirDelta: Float = 16f,
  val headTurnHoldMs: Long = 80L,
  val nodDownDelta: Float = 3f,
  val nodReturnFraction: Float = 0.85f,
  val nodReturnMaxDelta: Float = 12f,
  // Real blinks easily hit 0.7+. Sustained squinting / glare / hooded eyelids hover at 0.25–0.40 —
  // keep the "closed" threshold well above that band so passive eye states don't latch CLOSED.
  val blinkClosedThreshold: Float = 0.50f,
  val blinkOpenThreshold: Float = 0.25f,
  val earClosedThreshold: Float = 0.20f,
  val earOpenThreshold: Float = 0.25f,
  /** Eyes must be CLOSED for at least this long before a blink counts — filters out single-frame noise. */
  val blinkMinClosedMs: Long = 60L,
  val blinkMaxDurationMs: Long = 4000L,
  val mouthOpenThreshold: Float = 0.20f,
  val mouthOpenMarThreshold: Float = 0.20f,
  val mouthHoldMs: Long = 50L,
  val maxYawDuringBlink: Float = 30f,
  val maxPitchDuringBlink: Float = 30f,
  val maxYawDuringNod: Float = 40f,
  val maxYawDuringMouth: Float = 35f,
  val maxPitchDuringMouth: Float = 35f,
  val ovalCx: Float = 0.50f,
  val ovalCy: Float = 0.42f,
  val ovalRx: Float = 0.32f,
  val ovalRy: Float = 0.40f,
  val minFaceSize: Float = 0.10f,
  val maxFaceSize: Float = 0.62f,
  val captureDelayMs: Long = 700L,
  val captureMaxAttempts: Int = 90,
  val captureMaxYaw: Float = 18f,
  val captureMaxPitch: Float = 18f,
  val captureMaxMouthScore: Float = 0.20f,
  val captureMaxBlinkScore: Float = 0.25f,
  val captureMinEar: Float = 0.22f,
  val captureMaxMar: Float = 0.22f,
  val shuffleSteps: Boolean = true,
  /**
   * Subset of challenges to run, e.g. `listOf("nod", "blink", "mouth")`. Keys are
   * `"left"`, `"blink"`, `"right"`, `"nod"`, `"mouth"`. `null` or empty = all 5.
   * Unknown keys are dropped; duplicates de-duped. See [LivenessStep.resolve].
   */
  val steps: List<String>? = null,
  val faceDetectionConfidence: Float = 0.5f,
  val facePresenceConfidence: Float = 0.5f,
  val faceTrackingConfidence: Float = 0.5f,
  val cdnMaxRetries: Int = 5,
  val cdnAttemptTimeoutMs: Long = 45000L,
  val connectivityCheckTimeoutMs: Long = 5000L,

  // ── UI: shape, theme, layout (mirrors @daboss2003/liveness-web options) ───
  /** "oval" (default) or "circle". */
  val shape: String = "oval",
  /** When false, SDK-rendered instruction text, position hint, gesture icon and step dots are hidden. Sounds still play. */
  val showInstructions: Boolean = true,
  /** Minimum diameter (dp) of the visible face frame, so it stays usable on small screens. 0 = no floor. */
  val minSizeDp: Float = 220f,
  /** Ring stroke colour while face is in position. */
  val progressColor: Int = Color.parseColor("#12c95c"),
  /** Ring stroke colour while face is out of position. */
  val progressErrorColor: Int = Color.parseColor("#ff3b3b"),
  /** Ring stroke width in dp. */
  val progressWidthDp: Float = 3.5f,
  /** "round", "square" or "butt". */
  val progressLineCap: String = "round",
  /** Mask fill colour outside the cutout while face is in position. ARGB int. */
  val overlayColor: Int = 0xD1000000.toInt(),
  /** Mask fill colour outside the cutout while face is out of position. */
  val overlayErrorColor: Int = 0x8C000000.toInt(),
)
